src/main.rs

```diff
@@ -1,91 +1,186 @@
+        &self,
+        ctx: &Context,
+        guild: &Guild,
+    ) -> Result<ApplicationCommand, Box<dyn Error>>;
+
+    /// Called to handle an incoming application command with a name that
+    /// matches this impl's [`get_name`].
+    ///
+    /// This method must send some sort of response to the given
+    /// [`ApplicationCommandInteraction`], using
+    /// [`create_interaction_response`].
+    ///
+    /// # Errors
+    ///
+    /// This method returns an error if there is any issue handling this
+    /// application command. This may be due to the provided application
+    /// command information having unexpected arguments, an issue sending a
+    /// response to the application command, or any other issue specific to a
+    /// given application command.
+    ///
+    /// [`get_name`]: Self::get_name
+    /// [`create_interaction_response`]:
+    /// ApplicationCommandInteraction::create_interaction_response
+    async fn handle_command(
+        &self,
+        ctx: &Context,
+        command: ApplicationCommandInteraction,
+    ) -> Result<(), Box<dyn Error>>;
+
+    /// Gets the name of the application command associated with this impl. This
+    /// function must always return the same value for any given concrete impl.
+    fn get_name(&self) -> String;
+}
 
-/// Handler for the application - will implement an EventHandler as _part_ of this
 struct Handler;
 
 #[async_trait]
 impl EventHandler for Handler {
-    // From docs:
-    // Dispatched upon startup.
-    // Provides data about the bot and the guilds it’s in.
-    async fn ready(&self, ctx: Context, _data_about_bot: Ready) {
-        // Register each command name and description
-        let commands = ApplicationCommand::set_global_application_commands(&ctx.http, |commands| {
-            commands.create_application_command(|command| {
-                command.name("ping").description("A ping command")
-            })
-        })
-        .await;
-
-        println!(
-            "I now have the following global slash commands: {:#?}",
-            commands
-        );
-    }
-
-    // From docs:
-    // Dispatched when an interaction is created (e.g a slash command was used or a button was clicked).
-    // Provides the created interaction.
+    // Dispatched when an interaction is created (e.g. a slash command was used or a
+    // button was clicked). Provides the created interaction.
+    //
+    // When a new interaction is created, we check its name against all registered
+    // application commands (in [`APPLICATION_COMMANDS`]). If any match,
+    // we dispatch that application command's handler.
+    //
+    // # Panics
+    //
+    // Panics if none of the available [`APPLICATION_COMMANDS`] match, or if we run
+    // into an error handling this application command.
     async fn interaction_create(&self, ctx: Context, interaction: Interaction) {
-        // Figure out what interaction was sent
+        println!("{:#?}", interaction);
```

Comment:
Could probably do with a few more comments around here, and without this `println`.

Reply:
Should we implement formal logging on some level, instead of print statements?